### PR TITLE
mpi4.1: add split_type MPI_COMM_TYPE_RESOURCE_GUIDED

### LINKED ISSUE #6763: mpi4.1: MPI_Get_hw_resource_info
https://github.com/mpi-forum/mpi-issues/issues/538
https://github.com/mpi-forum/mpi-standard/issues/761
https://github.com/mpi-forum/mpi-standard/issues/762

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description
Add `MPI_Get_hw_resource_info` and new split_type `MPI_COMM_TYPE_RESOURCE_GUIDED`.
Refactor code `MPII_collect_info_key` that retrieves info key and collectively verifies they are equal. Internally use `MPIR_Allreduce_equal`.

[skip warnings]

Fixes #6763 
## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by dalcinl** [comment]
When you have a chance, could you please rebase this PR to latest main?

**Comment by hzhou** [comment]
> When you have a chance, could you please rebase this PR to latest main?

done

**Comment by hzhou** [comment]
test:mpich/ch3/tcp
test:mpich/ch4/ofi

**Comment by dalcinl** [comment]
@hzhou mpi4py tests are segfaulting within `MPI_Comm_split_type`. It's late here, I'll continue tomorrow.

**Comment by hzhou** [comment]
> @hzhou mpi4py tests are segfaulting within `MPI_Comm_split_type`. It's late here, I'll continue tomorrow.

Thanks. I believe we are also seeing that in our tests. Will fix.

**Comment by hzhou** [comment]
test:mpich/ch3/tcp
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

File: src/mpi/comm/comm_split_type.c [hunk 1]
```diff
@@ -40,11 +42,41 @@ int MPIR_Comm_split_type(MPIR_Comm * user_comm_ptr, int split_type, int key,
...
         MPIR_ERR_CHECK(mpi_errno);
+    } else if (split_type == MPI_COMM_TYPE_RESOURCE_GUIDED) {
+        /* similar to MPI_COMM_TYPE_HW_GUIDED, but may also use process set name to split */
+        /* first make sure we have consistent info keys */
+        const char *resource_type = NULL;
+        const char *pset_name = NULL;
+        mpi_errno = MPII_collect_info_key(comm_ptr, info_ptr, "mpi_hw_resource_type",
+                                          &resource_type);
+        MPIR_ERR_CHECK(mpi_errno);
+        mpi_errno = MPII_collect_info_key(comm_ptr, info_ptr, "mpi_pset_name", &resource_type);
```
Comment (raffenet) [inline src/mpi/comm/comm_split_type.c:68]: Should last argument be `&pset_name`?

Comment (hzhou) [reply src/mpi/comm/comm_split_type.c]: Good catch! Just fixed.

**Review by raffenet — APPROVED** [review]

**Pull request merged**